pleskdistup/common/src/util.py

```diff
@@ -9,31 +9,62 @@
 # Returns standard output
 def logged_check_call(cmd: typing.Union[typing.Sequence[str], str], **kwargs) -> str:
     log.info(f"Running: {cmd!r}. Output:")
+    stdout = []
+
+    def proc_stdout(line: str) -> None:
+        stdout.append(line)
+
+    def proc_stderr(line: str) -> None:
+        log.err(line)
+
+    exit_code = exec_get_output_streamed(cmd, proc_stdout, proc_stderr, **kwargs)
+    if exit_code != 0:
+        log.err(f"Command {cmd!r} failed with return code {exit_code}")
+        raise subprocess.CalledProcessError(returncode=exit_code, cmd=cmd, output="\n".join(stdout))
+
+    log.info(f"Command {cmd!r} finished successfully")
+    return "\n".join(stdout)
 
-    # I beleive we should be able pass argument to the subprocess function
+
+# Allows to get stdout/stderr by streaming line by line, calling callbacks
+# and returns process exit code
+def exec_get_output_streamed(
+    cmd: typing.Union[typing.Sequence[str], str],
+    process_stdout_line: typing.Optional[typing.Callable[[str], None]],
+    process_stderr_line: typing.Optional[typing.Callable[[str], None]],
+    **kwargs,
+) -> int:
+    # I believe we should be able pass argument to the subprocess function
     # from the caller. So we have to inject stdout/stderr/universal_newlines
-    kwargs["stdout"] = subprocess.PIPE
-    kwargs["stderr"] = subprocess.STDOUT
+    if process_stdout_line is None:
+        kwargs["stdout"] = subprocess.DEVNULL
+    else:
+        kwargs["stdout"] = subprocess.PIPE
+    if process_stderr_line is None:
+        kwargs["stderr"] = subprocess.DEVNULL
+    else:
+        kwargs["stderr"] = subprocess.PIPE
     kwargs["universal_newlines"] = True
 
-    stdout = []
     process = subprocess.Popen(cmd, **kwargs)
-    while None is process.poll():
-        if not process.stdout:
-            log.err(f"Can't get process output from {cmd!r}")
-            raise RuntimeError(f"Can't get process output from {cmd!r}")
-        line = process.stdout.readline()
-        if line:
-            stdout.append(line)
-            if line.strip():
-                log.info(line.strip(), to_stream=False)
-
-    if process.returncode != 0:
-        log.err(f"Command {cmd!r} failed with return code {process.returncode}")
-        raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd, output="\n".join(stdout))
+    if process_stdout_line is None and process_stderr_line is None:
+        process.communicate()
+        return process.returncode
```

Comment:
Now this doesn't raise on non-zero result. This is a significant behavior change.

Reply:
Behavior change could only be about previous function `logged_check_call()` which already [raises when exit_code is nonzero](https://github.com/plesk/dist-upgrader/pull/82/files#diff-7da365a7435ef4d10f5c4359e026a138b569ad1448f15af07a6166e72c07d9a9R21)

Reply:
Just to give more detail:
`logged_check_call` utilizes `exec_get_output_streamed` and its behavior should be intact. However we use `exec_get_output_streamed` to retry transient failures in dpkg* commands.